### PR TITLE
WIP: Migrate Travis jobs to GitHub Actions

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
This PR follows https://github.com/mne-tools/mne-python/pull/8335#issuecomment-726761191 and migrates all remaining jobs on Travis to GitHub Actions.

ToDo:

- [x] Migrate Travis jobs
- [x] Remove travis conf file
- [x]  Try separate jobs to fix `compat / old` job
- [x] Add inline annotations for codespell and flake (suggested in https://github.com/mne-tools/mne-python/pull/8335#issuecomment-726761191)
- [x] Remove `TRAVIS_OS_NAME` variable
- [x] Add `[skip azp]` mechanism (suggested in https://github.com/mne-tools/mne-python/pull/8514#issuecomment-726797122)

This is still work in progress.

### CODE REVIEW

**Comment by larsoner** [comment]
While you're digging around in CIs, could you add a step to all Azure runs that actually checks the commit message for `[skip azp]`, `[azp skip]`, and `[ci skip]` and just exit the run with succeeded status if detected? I've been hoping for months that Azure would add native support for this but they haven't.

I see two ways of doing it:

1. Reuse a bit of code in each job, which you could adapt from SciPy here:

    https://github.com/scipy/scipy/blob/master/.circleci/config.yml#L11

2. Make a "check skip" job that all other jobs depend on, and only run the other jobs if the check skip job says to do so. I don't *think* this will bottleneck things much since this job would run very quickly, then all N other Azure jobs would spawn.

Not sure which is better, or if there is another better alternative

**Comment by GuillaumeFavelier** [comment]
When do the following variables need to be exposed? I mean in which `.sh` script they're used

* pip/pre
```
            MNE_STIM_CHANNEL: 'STI101'
            MNE_LOGGING_LEVEL: 'info'
```

* minimal
```
            MNE_DONTWRITE_HOME: true
            MNE_FORCE_SERIAL: true
            MNE_SKIP_NETWORK_TEST: 1
            MNE_SKIP_TESTING_DATASET_TESTS: true
```

**Comment by GuillaumeFavelier** [comment]
> When do the following variables need to be exposed? 

Nevermind, I exposed everything with the state of the art syntax I had no idea existed:

https://github.community/t/support-global-environment-variables/16146

Now, I have to make miniconda happy (conda activate fails) and ensure that MacOS is still okay.

**Comment by GuillaumeFavelier** [comment]
I could split the Linux jobs in 2 workflows `Linux` (`pip` and `conda`) and `Linux / Compat` (`old` and `minimal`).

**Comment by GuillaumeFavelier** [comment]
`linux/old` [failed](https://github.com/mne-tools/mne-python/pull/8514/checks?check_run_id=1397531341#step:11:2415) with:

```
=================================== FAILURES ===================================
__________________________ test_xdawn_regularization ___________________________
mne/preprocessing/tests/test_xdawn.py:201: in test_xdawn_regularization
    xd.fit(epochs)
E   Failed: DID NOT RAISE <class 'ValueError'>
```

`linux/conda` [had a segfault]( https://github.com/mne-tools/mne-python/pull/8514/checks?check_run_id=1397531341#step:11:2415) around `mne/forward/tests/test_make_forward.py::test_make_forward_solution`

<details>

```
mne/forward/tests/test_make_forward.py::test_make_forward_solution_kit SKIPPED [ 21%]
Fatal Python error: Segmentation fault

Thread 0x00007f368a1c1700 (most recent call first):
  File "/usr/share/miniconda/envs/mne/lib/python3.8/threading.py", line 306 in wait
  File "/usr/share/miniconda/envs/mne/lib/python3.8/threading.py", line 558 in wait
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/tqdm/_monitor.py", line 59 in run
  File "/usr/share/miniconda/envs/mne/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/share/miniconda/envs/mne/lib/python3.8/threading.py", line 890 in _bootstrap

Current thread 0x00007f36fa5c7740 (most recent call first):
  File "/home/runner/work/mne-python/mne-python/mne/surface.py", line 319 in _project_onto_surface
  File "/home/runner/work/mne-python/mne-python/mne/forward/_compute_forward.py", line 253 in _bem_specify_els
  File "/home/runner/work/mne-python/mne-python/mne/forward/_compute_forward.py", line 788 in _prep_field_computation
  File "<decorator-gen-288>", line 24 in _prep_field_computation
  File "/home/runner/work/mne-python/mne-python/mne/forward/_compute_forward.py", line 918 in _compute_forwards
  File "<decorator-gen-289>", line 24 in _compute_forwards
  File "/home/runner/work/mne-python/mne-python/mne/forward/_make_forward.py", line 603 in make_forward_solution
  File "<decorator-gen-295>", line 24 in make_forward_solution
  File "/home/runner/work/mne-python/mne-python/mne/forward/tests/test_make_forward.py", line 224 in test_make_forward_solution
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/python.py", line 184 in pytest_pyfunc_call
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/python.py", line 1627 in runtest
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 163 in pytest_runtest_call
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 256 in <lambda>
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 310 in from_call
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 255 in call_runtest_hook
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 216 in call_and_report
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 127 in runtestprotocol
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/runner.py", line 110 in pytest_runtest_protocol
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/main.py", line 338 in pytest_runtestloop
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/main.py", line 313 in _main
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/main.py", line 257 in wrap_session
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/main.py", line 306 in pytest_cmdline_main
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/config/__init__.py", line 164 in main
  File "/usr/share/miniconda/envs/mne/lib/python3.8/site-packages/_pytest/config/__init__.py", line 187 in console_main
  File "/usr/share/miniconda/envs/mne/bin/pytest", line 8 in <module>
./tools/github_actions_test.sh: line 28:  4718 Segmentation fault      (core dumped) pytest -m "${CONDITION}" --tb=short --cov=mne --cov-report xml -vv 
```
</details>

**Comment by GuillaumeFavelier** [comment]
Surpringly, removing cruft (c400c3a) seems to solve 2). I can only assume that one of the variables I removed had an influence even though it was not set:

```py
      CONDA_DEPENDENCIES: ${{ matrix.CONDA_DEPENDENCIES }}
      DEPS: ${{ matrix.DEPS }}
      MNE_DONTWRITE_HOME: ${{ matrix.MNE_DONTWRITE_HOME }}
      MNE_FORCE_SERIAL: ${{ matrix.FORCE_SERIAL }}
      MNE_SKIP_NETWORK_TEST: ${{ matrix.MNE_SKIP_NETWORK_TEST }}
      MNE_SKIP_TESTING_DATASET_TESTS: ${{ matrix.MNE_SKIP_TESTING_DATASET_TESTS }}
```

And 1) seems random to me.

I restarted the workflows to confirm.

---

I wanted to work on test matrices to have more compact configs but if it's not stable enough, the safe route is just to define each job separately.

**Comment by GuillaumeFavelier** [comment]
Is it good enough as annotation?

image


I have to click on `Check failure` to have more infos. I thought it would be displayed inline automatically.

image

**Comment by GuillaumeFavelier** [comment]
For a full list of the [codespell action features](https://github.com/codespell-project/actions-codespell), I think looking at the code is better:

https://github.com/codespell-project/actions-codespell/blob/master/action.yml

Also, `usage` is not available apparently as builtin:

```
  --builtin BUILTIN-LIST
                        Comma-separated list of builtin dictionaries to
                        include (when "-D -" or no "-D" is passed). Current
                        options are: 'clear' for unambiguous errors, 'rare'
                        for rare but valid words, 'informal' for informal
                        words, 'code' for words common to code and/or
                        mathematics, 'names' for valid proper names that might
                        be typos, 'en-GB_to_en-US' for corrections from en-GB
                        to en-US. The default is "--builtin clear,rare".
```

I got the same message locally and I had to install `codespell` from `master` to fix it. I suppose it may become available in the future.

To move forward with this, I will remove the 'usage' builtin from the action config.

**Comment by GuillaumeFavelier** [comment]
It would be nice to have access to the quiet level as well (our default is `-q 3`):

```
  -q QUIET_LEVEL, --quiet-level QUIET_LEVEL
                        bitmask that allows suppressing messages:
                        - 0: print all messages.
                        - 1: disable warnings about wrong encoding.
                        - 2: disable warnings about binary files.
                        - 4: omit warnings about automatic fixes that were disabled in the dictionary.
                        - 8: don't print anything for non-automatic fixes.
                        - 16: don't print the list of fixed files.
                        As usual with bitmasks, these levels can be combined; e.g. use 3 for levels 1+2, 7 for 1+2+4, 23 for 1+2+4+16, etc. The default mask is 2.
```

Because the current annotations are not really useful (about 10 like that):

image

But since we use the `@master` version of the action, we can expect to get the changes as soon as they are merged.

**Comment by larsoner** [comment]
Feel free to open a PR to codespell

**Comment by GuillaumeFavelier** [comment]
It seems skipping azure was requested since https://github.com/microsoft/azure-pipelines-agent/issues/858. It ended up being implemented for [CI builds](https://github.com/microsoft/azure-pipelines-agent/issues/858#issuecomment-475768046) but [not for PRs](https://github.com/microsoft/azure-pipelines-agent/issues/858#issuecomment-500874884).

I tried [the condition described in the thread](https://github.com/microsoft/azure-pipelines-agent/issues/858#issuecomment-457027046) in 5f5c619:

```
steps:
  - bash: echo "Run tests"
    condition: not(contains(variables['Build.SourceVersionMessage'], '[skip ci]'))
```

But the variable `Build.SourceVersionMessage` is something like:

`Merge 5f5c619fdb98d21331a1314bc036417cf45070d2 into 5d43084ffc12ae4c8d73c5db991856ea8e2e47e5`

So the condition in it's current state is not working. Plus, apparently, it needs to be repeated at each step to be able to skip the full job.

I'll continue reading documentation, there is maybe a better variable for that.

**Comment by GuillaumeFavelier** [comment]
Same old failure on `compat / old` for `Xdawn`:

Here is the config:

- GitHub Actions:

```
Platform:      Linux-5.4.0-1031-azure-x86_64-with-debian-buster-sid
Python:        3.6.12 |Anaconda, Inc.| (default, Sep  8 2020, 23:10:56)  [GCC 7.3.0]
Executable:    /usr/share/miniconda/envs/test/bin/python
CPU:           x86_64: 2 cores
Memory:        Unavailable (requires "psutil" package)
mne:           0.22.dev0
numpy:         1.15.4 {blas=mkl_rt, lapack=mkl_rt}
scipy:         1.1.0
matplotlib:    3.0.3 {backend=Qt5Agg}

sklearn:       0.20.2
numba:         Not found
nibabel:       3.2.0
cupy:          Not found
pandas:        0.23.4
dipy:          Not found
mayavi:        Not found
pyvista:       Not found
vtk:           Not found
mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/usr/share/miniconda/envs/test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/usr/share/miniconda/envs/test/include']
blas_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/usr/share/miniconda/envs/test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/usr/share/miniconda/envs/test/include']
blas_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/usr/share/miniconda/envs/test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/usr/share/miniconda/envs/test/include']
lapack_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/usr/share/miniconda/envs/test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/usr/share/miniconda/envs/test/include']
lapack_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/usr/share/miniconda/envs/test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/usr/share/miniconda/envs/test/include']
```

- Travis:

```
Platform:      Linux-5.3.0-1029-gcp-x86_64-with-debian-buster-sid
Python:        3.6.12 |Anaconda, Inc.| (default, Sep  8 2020, 23:10:56)  [GCC 7.3.0]
Executable:    /home/travis/miniconda/envs/test/bin/python
CPU:           x86_64: 2 cores
Memory:        Unavailable (requires "psutil" package)
mne:           0.22.dev0
numpy:         1.15.4 {blas=mkl_rt, lapack=mkl_rt}
scipy:         1.1.0
matplotlib:    3.0.3 {backend=agg}

sklearn:       0.20.2
numba:         Not found
nibabel:       3.2.0
cupy:          Not found
pandas:        0.23.4
dipy:          Not found
mayavi:        Not found
pyvista:       Not found
vtk:           Not found
mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/travis/miniconda/envs/test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/travis/miniconda/envs/test/include']
blas_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/travis/miniconda/envs/test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/travis/miniconda/envs/test/include']
blas_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/travis/miniconda/envs/test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/travis/miniconda/envs/test/include']
lapack_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/travis/miniconda/envs/test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/travis/miniconda/envs/test/include']
lapack_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/travis/miniconda/envs/test/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/travis/miniconda/envs/test/include']
```

**Comment by larsoner** [comment]
You should be able to do some gymnastics with `git log HEAD~2..HEAD~1` or so followed by [defining your own variable](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#understand-variable-syntax)

**Comment by larsoner** [comment]
> Same old failure on compat / old for Xdawn:

Looks okay now, probably because the Xdawn fixes have been merged into `master` by now

**Comment by larsoner** [comment]
... and keep in mind you'll want to check for any one of `[skip azp]`, `[azp skip]`, or `[ci skip]` probably (feel free to add `[skip ci]` if you want, but please also add this in the GitHub actions skip list, too)

**Comment by GuillaumeFavelier** [comment]
Finally a skip. Now I have to work on a reliable conditional

**Comment by GuillaumeFavelier** [comment]
Okay I think Azure is good to go now

File: .github/workflows/compat.yml [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        run: ./tools/github_actions_locale.sh
+        name: 'Print locale'
+      - shell: bash -el {0}
+        run: ./tools/github_actions_test.sh
+        name: 'Run tests'
+      - uses: codecov/codecov-action@v1
+        if: success()
+        name: 'Upload coverage to CodeCov'
+
+  # Minimal (runs with and without testing data)
```
Comment (larsoner) [inline .github/workflows/compat.yml:61]: Separate files would be better for restartability

Comment (GuillaumeFavelier) [reply .github/workflows/compat.yml:61]: I'll separate them all then

File: .github/workflows/linux.yml [hunk 2]
```diff
@@ -59,3 +59,53 @@ jobs:
       - uses: codecov/codecov-action@v1
         if: success()
         name: 'Upload coverage to CodeCov'
+
+  # PIP + non-default stim channel + log level info
+  pip:
```
Comment (larsoner) [inline .github/workflows/linux.yml:64]: same

File: .github/workflows/maint.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
+name: 'maint'
```
Comment (larsoner) [inline .github/workflows/maint.yml:1]: Maybe `codespell_and_flake` ?

File: azure-pipelines.yml [hunk 4]
```diff
@@ -10,240 +10,262 @@ pr:
...
+    - bash: |
+        set -e
+        python -m pip install --upgrade pip setuptools
+        python -m pip install numpy scipy matplotlib -r requirements_testing.txt
+      displayName: Install dependencies
+      condition: always()
+    - bash: |
+        make flake
+      displayName: make flake
+      condition: always()
```
Comment (larsoner) [inline azure-pipelines.yml:59]:
```suggestion
```

File: azure-pipelines.yml [hunk 5]
```diff
@@ -10,240 +10,262 @@ pr:
...
+    - bash: |
+        make pydocstyle
+      displayName: make pydocstyle
+      condition: always()
+    - bash: |
+        make codespell-error
+      displayName: make codespell-error
+      condition: always()
```
Comment (larsoner) [inline azure-pipelines.yml:67]:
```suggestion
```

File: .github/workflows/compat.yml (hunk 1 above)
Comment (larsoner) [reply .github/workflows/compat.yml:61]: Yes the GitHub ones (Azure can and should stay a single file)

**Comment by GuillaumeFavelier** [comment]
For future reference, I got the infos of the correct usage of `stage`& `dependencies` from:

https://github.com/microsoft/azure-pipelines-tasks/issues/4743#issuecomment-643469964

**Comment by GuillaumeFavelier** [comment]
This is ready for reviews but:

1) I still have doubts about `compat / old` job (it failed in baae1d1 for example). The failure seems random and I don't notice any big differences compared to the Travis environment.

2) The GitHub actions don't react to `[skip ci]`, `[ci skip]` ...etc. And this could be fixed in another PR

**Comment by larsoner** [comment]
@GuillaumeFavelier our minimum is 1.15 and that part of the test is skipped for 1.16+, we should just skip it entirely. For now maybe just comment it out so that someday we come back to it?

**Comment by larsoner** [comment]
I'll merge and we can iterate more as necessary, thanks @GuillaumeFavelier !

**Comment by agramfort** [comment]
can you update the badges on readme @GuillaumeFavelier ? thanks heaps !

**Comment by larsoner** [comment]
Looks like you can make them for GitHub actions like https://github.com/mne-tools/mne-python/workflows/linux%20/%20conda/badge.svg?branch=master:

**Comment by cbrnr** [comment]
Great work! Do we have any numbers on how long tests take on GitHub Actions vs. how long they took on Travis? AFAIK the longest job on Travis was around 50-55 minutes...

**Comment by larsoner** [comment]
Yes the `pip pre` linux on Travis was 50-55, looks like on the last green check mark above this job was ~40 minutes

**Pull request merged**